### PR TITLE
[REEF-1354] Yarn bootstrap for c# multi-runtime client

### PR DESCRIPTION
**Pull request by shulmanb**
This addresses the issue by:
1) Adding bootstrap mechanism for multiruntime

JIRA:
  [REEF-1354](https://issues.apache.org/jira/browse/REEF-1354)

### CODE REVIEW

**Comment by shulmanb** [comment]
@afchung @markusweimer This is the java bootstrap part from REEF-1054.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+  static AvroMultiRuntimeAppSubmissionParameters readMultiRuntimeAppSubmissionParametersFromFile(final File file)
+      throws IOException {
+    try (final FileInputStream fileInputStream = new FileInputStream(file)) {
+      // This is mainly a test hook.
+      return readMultiRuntimeAppSubmissionParametersFromInputStream(fileInputStream);
+    }
+  }
+
+  static AvroMultiRuntimeAppSubmissionParameters readMultiRuntimeAppSubmissionParametersFromInputStream(
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: This was previously added to expose for unit tests. Do we have any unit tests here? If not, please add one.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            multiruntimeAppSubmissionParams.getDefaultRuntimeName().toString());
+
+    // generate multi runtime configuration
+    ConfigurationModule multiRuntimeDriverConfiguration = MultiRuntimeDriverConfiguration.CONF;
+    for (final CharSequence runtimeName : multiruntimeAppSubmissionParams.getRuntimes()){
+      multiRuntimeDriverConfiguration = multiRuntimeDriverConfiguration.set(
+              MultiRuntimeDriverConfiguration.RUNTIME_NAMES, runtimeName.toString());
+    }
+
+    multiRuntimeDriverConfiguration = multiRuntimeDriverConfiguration
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: `set` at declaration.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    // generate multi runtime definition
+    final MultiRuntimeDefinitionBuilder multiRuntimeDefinitionBuilder = new MultiRuntimeDefinitionBuilder();
+
+    if (multiruntimeAppSubmissionParams.getLocalRuntimeAppParameters() != null){
+      addLocalRuntimeDefinition(
+              multiruntimeAppSubmissionParams.getLocalRuntimeAppParameters(),
+              jobSubmissionParameters, multiRuntimeDefinitionBuilder);
+    }
+
+    if (multiruntimeAppSubmissionParams.getYarnRuntimeAppParameters() != null){
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Sorry, can you remind me how this can be a valid use-case? At least one of them have to be non-null right?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 2 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Done.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 3 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: At least one of them has to be non-null. Will add a validation for both nulls case.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 1 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Ok. Will do.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 3 above)
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: So how is this a valid use-case?

**Comment by afchung** [comment]
@shulmanb done with a first pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 3 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: You can submit a yarn job that will only use local runtime for containers.

Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: My main suspicion rises from how the YARN Driver communication is set up with the RM if this is, in the following cases:
1. Bound (`MultiRuntimeDefinitionBuilder` seems to also bind it, will it be bound twice?)
2. Not Bound (if the communication will be properly set up at all)

Sorry, I wasn't entirely involved in the last PR. @markusweimer could you comment on this as well?

Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: What is bound to REEF runtime is the multiruntime:
public static final ConfigurationModule CONF = new MultiRuntimeDriverConfiguration()
      .bindImplementation(ResourceLaunchHandler.class, MultiRuntimeResourceLaunchHandler.class)
      .bindImplementation(ResourceRequestHandler.class, MultiRuntimeResourceRequestHandler.class)
      .bindImplementation(ResourceReleaseHandler.class, MultiRuntimeResourceReleaseHandler.class)
      .bindImplementation(ResourceManagerStartHandler.class, MultiRuntimeResourceManagerStartHandler.class)
      .bindImplementation(ResourceManagerStopHandler.class, MultiRuntimeResourceManagerStopHandler.class)
      .bindNamedParameter(SerializedRuntimeDefinition.class, SERIALIZED_RUNTIME_DEFINITION)
      .bindNamedParameter(LaunchID.class, JOB_IDENTIFIER)
      .bindNamedParameter(JobIdentifier.class, JOB_IDENTIFIER)
      .bindNamedParameter(ClientRemoteIdentifier.class, CLIENT_REMOTE_IDENTIFIER)
      .bindSetEntry(DefinedRuntimes.class, RUNTIME_NAMES)

I don't think we need to launch any other YArn handlers if we are not actually requesting Yarn containers.

I do bind the following:

        .bindNamedParameter(JobSubmissionDirectoryPrefix.class,
            yarnJobSubmissionParams.getJobSubmissionDirectoryPrefix().toString())
        .bindImplementation(RuntimeClasspathProvider.class, YarnClasspathProvider.class)
        .bindConstructor(YarnConfiguration.class, YarnConfigurationConstructor.class)

Yarn job submission parameters are read in as a JObParams not App params. I will test it again before merge on YArRest/HDI

**Comment by shulmanb** [comment]
@afchung @markusweimer I pushed new version.I tried testing only with Local runtime or only with Yarn and it does not work (works only with both) as avro types are not nullable by default and you need union types to represent nullable types. I tried changing it to union types but apparently c# avro for nullable types is not readable by java side. For example it produces:

{"sharedAppSubmissionParameters":{"tcpBeginPort":8900,"tcpRangeCount":1000,"tcpTryCount":1000},"localRuntimeAppParameters":{"sharedAppSubmissionParameters":{"tcpBeginPort":8900,"tcpRangeCount":1000,"tcpTryCount":1000},"maxNumberOfConcurrentEvaluators":2},"yarnRuntimeAppParameters":null,"defaultRuntimeName":"Local","runtimes":["Local"]}

When the expected is:
{"sharedAppSubmissionParameters":{"tcpBeginPort":8900,"tcpRangeCount":1000,"tcpTryCount":1000},
"localRuntimeAppParameters":{"org.apache.reef.reef.bridge.client.avro.AvroYarnAppSubmissionParameters":{"sharedAppSubmissionParameters":{"tcpBeginPort":8900,"tcpRangeCount":1000,"tcpTryCount":1000}},"maxNumberOfConcurrentEvaluators":2},
"yarnRuntimeAppParameters":null,"defaultRuntimeName":"Local","runtimes":["Local"]}

Any ideas?

**Comment by markusweimer** [comment]
A hacky workaround would be to use an array for this. When it is empty, it means that the item is not bound. If its size is bigger than 1, we bail. It is a hack, but could work.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 3 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: @markusweimer @afchung I pushed a new version. I tested it on HDI with Yarn and Local, Local only and Yarn only. The test I ran is using HelloTaskMessage and saw that I receive messages from the expected runtimes. What is the best way to test HTTP endpoint?

**Comment by shulmanb** [comment]
@markusweimer @afchung I pushed a new version. I tested it on HDI with Yarn and Local, Local only and Yarn only. The test I ran is using HelloTaskMessage and saw that I receive messages from the expected runtimes. What is the best way to test HTTP endpoint?

**Comment by shulmanb** [comment]
@markusweimer I can think of a more generic way configuring multiple runtimes (without the need to actually code method and check for existence of each one of them), but would prefer to have that in , cut 0.15 and then continue, as we are not exposing it as a public c# to anyone at this stage.
I can mark this particular class as unstable.

**Comment by shulmanb** [comment]
Closing this PR. Will reopen after PID issue for c# .NET is fixed.

**Comment by shulmanb** [comment]
@afchung, @markusweimer  I reopened this JIRA, please take a look.

**Comment by shulmanb** [comment]
@afchung You were right, I actually need yarn runtime to be initialized even if I don't submit evaluators to it. Added that.

**Comment by shulmanb** [comment]
@afchung, @markusweimer Please take a look.

**Comment by markusweimer** [comment]
This is blocked by #983, right?

**Comment by shulmanb** [comment]
@markusweimer No. It is not. For bootstrapping the driver config generated differently in the bootstrapping code directly. In #983 the bug is on the client side.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                    jobSubmissionParameters.getJobId().toString())
+            .set(YarnDriverConfiguration.CLIENT_REMOTE_IDENTIFIER,
+                    ClientRemoteIdentifier.NONE)
+            .set(YarnDriverConfiguration.JVM_HEAP_SLACK, 0.0)
+            .set(YarnDriverConfiguration.RUNTIME_NAMES, RuntimeIdentifier.RUNTIME_NAME)
+            .build();
+  }
+
+  /**
+   * Adds yarn runtime definitions to the builder, with a dumy name, to prevent evaluator submission.
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: `dummy`

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+   */
+  private void addDummyYarnRuntimeDefinition(
+          final AvroYarnJobSubmissionParameters yarnJobSubmissionParams,
+          final AvroJobSubmissionParameters jobSubmissionParameters,
+          final MultiRuntimeDefinitionBuilder builder) {
+    // create and serialize yarn configuration if defined
+    final Configuration yarnDriverConfiguration =
+            createYarnConfiguration(yarnJobSubmissionParams, jobSubmissionParameters);
+    // add yarn runtime to the builder
+    builder.addRuntime(yarnDriverConfiguration, DUMMY_YARN_RUNTIME);
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Can you remind me what submission looks like? How does this prevent submission to the YARN environment?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    return reader.read(null, decoder);
+  }
+
+  /**
+   * Writes the driver configuration files to the provided location.
+   * @param bootstrapJobArgsLocation The path fro the job args file
+   * @param bootstrapAppArgsLocation The path for the app args file
+   * @throws IOException
+   */
+  public String writeDriverConfigurationFile(final String bootstrapJobArgsLocation,
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Doesn't need to be public.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    final JsonDecoder decoder = DecoderFactory.get().jsonDecoder(
+            AvroMultiRuntimeAppSubmissionParameters.getClassSchema(), inputStream);
+    final SpecificDatumReader<AvroMultiRuntimeAppSubmissionParameters> reader = new SpecificDatumReader<>(
+            AvroMultiRuntimeAppSubmissionParameters.class);
+    return reader.read(null, decoder);
+  }
+
+  /**
+   * Writes the driver configuration files to the provided location.
+   * @param bootstrapJobArgsLocation The path fro the job args file
```
Comment (afchung) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: `for`

**Comment by afchung** [comment]
Looks good to me overall, few minor comments.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 6 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Done.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 5 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: It prevents teh submission by binding this runtime with default identifier, so when evaluator for Yarn runtime is requested, RuntimesHost it won't be a valid runtime identifier

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 4 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Done

**Comment by shulmanb** [comment]
@afchung @markusweimer pushed new version. Can it be merged?

**Comment by afchung** [comment]
Looks good to me, will wait for @markusweimer to sign off. Thanks!

**Comment by shulmanb** [comment]
@markusweimer Can you please take a look?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+import javax.inject.Inject;
+import java.io.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * This is the Java Driver configuration generator for .NET Drivers that generates
+ * the Driver configuration at runtime for multiruntime. Called by {@link MultiRuntimeYarnBootstrapREEFLauncher}.
+ */
+final class MultiRuntimeYarnBootstrapDriverConfigGenerator {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Mark with `@DriverSide`

Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Wait, this is in the `.client` package. Where is it run?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+                    jobSubmissionParameters.getJobId().toString())
+            .set(YarnDriverConfiguration.CLIENT_REMOTE_IDENTIFIER,
+                    ClientRemoteIdentifier.NONE)
+            .set(YarnDriverConfiguration.JVM_HEAP_SLACK, 0.0)
+            .set(YarnDriverConfiguration.RUNTIME_NAMES, RuntimeIdentifier.RUNTIME_NAME)
+            .build();
+  }
+
+  /**
+   * Adds yarn runtime definitions to the builder, with a dummy name, to prevent evaluator submission.
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Why is this needed? The comment describes the *what*, but not the *why*.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+                              JobDriver.DriverRestartFailedEvaluatorHandler.class)
+                      .build();
+
+      return Configurations.merge(driverConfiguration, yarnDriverRestartConfiguration, driverRestartConfiguration);
+    }
+
+    return driverConfiguration;
+  }
+
+  private static AvroMultiRuntimeAppSubmissionParameters readMultiRuntimeAppSubmissionParametersFromFile(
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: This serialization code would be better placed into a injectable `*Serlializer` class like we do for `Configuration`. That way, we can reuse it in other places that need to deserialize it, e.g. the Driver.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * This is a bootstrap launcher for YARN for submission of multiruntime jobs from C#. It allows for Java Driver
+ * configuration generation directly on the Driver without need of Java dependency if REST
+ * submission is used. Note that the name of the class must contain "REEFLauncher" for the time
+ * being in order for the Interop code to discover the class.
+ */
+@Interop(CppFiles = "DriverLauncher.cpp")
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: This is in the `client` package, yet is part of the interop which only exists on the Driver. Please clarify.

**Comment by markusweimer** [comment]
I've done a pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 8 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: This bootstraps the driver. Not sure why this is a client package (all our bootstrap code is in this package). Renaming this should be a different JIRA as I am not sure how it affects current c# client.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 9 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Will add docs. This is needed to enforce usage of specified runtimes, so if user did not specify yarn runtime for evaluators we still need to instantiate one (as it registers with RM) but won't allow submitting evaluators to it.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java (hunk 11 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: All the bootstrap code (including existing) is in this package. The package name is clearly wrong (haven't noticed it before). We can create separate JIRA for it.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    return reader.read(null, decoder);
+  }
+
+  /**
+   * Writes the driver configuration files to the provided location.
+   * @param bootstrapJobArgsLocation The path for the job args file
+   * @param bootstrapAppArgsLocation The path for the app args file
+   * @throws IOException
+   */
+  String writeDriverConfigurationFile(final String bootstrapJobArgsLocation,
```
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Done.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 10 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: What serialization code you are referring to exactly?

**Comment by shulmanb** [comment]
@markusweimer addresses some comments and responded in the PR to questions

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 10 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: The one I left the comment on and all the other methods starting in `read` or `write`. Those are all concerned with serialization, and it would be good to bundle them in the spirit of the `org.apache.reef.tang.formats.AvroConfigurationSerializer`. That way, this functionality can be reused and tested more easily.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java (hunk 11 above)
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: I believe that is because the class does things for both the Driver and the Client. That seems dangerous from a maintainability perspective.

Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: I agree. But we should change that for all the bootstrap mechanims (this is no different from the current one). We should create a JIRA for this and work on this.

Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: +1 Can you please file the JIRA and link to it here from a `//TODO`?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs [hunk 13]
```diff
@@ -78,8 +78,8 @@ public void Submit(JobRequest jobRequest)
             // todo: Future client interface should be async.
             // Using GetAwaiter().GetResult() instead of .Result to avoid exception
             // getting wrapped in AggregateException.
-            var newApplication = _yarnRMClient.CreateNewApplicationAsync().GetAwaiter().GetResult();
-            string applicationId = newApplication.ApplicationId;
+            ////var newApplication = _yarnRMClient.CreateNewApplicationAsync().GetAwaiter().GetResult();
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: please remove the commented out code.

**Comment by markusweimer** [comment]
Thanks! There are still a couple of things where you mentioned you'd address them that aren't addressed in the most recent code. Can you have a look?

File: lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs (hunk 13 above)
Comment (shulmanb) [inline lang/cs/Org.Apache.REEF.Client/YARN/YarnREEFDotNetClient.cs]: This got it by mistake. No c# changes are required. Sorry.

**Comment by shulmanb** [comment]
@markusweimer Please take a look.

**Comment by markusweimer** [comment]
Doing a pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Searializer class for the AvroMultiRuntimeAppSubmissionParameters.
+ */
+final class AvroMultiRuntimeAppSubmissionParametersSerializer {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java]: Wouldn't we also expect `toFile()` methods in these?

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+  private AvroMultiRuntimeAppSubmissionParametersSerializer(){
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java]: An alternative is to make this class injectable, which makes it easier to mock different implementations during tests.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+
+
+  /**
+   * Writes the driver configuration files to the provided location.
+   * @param bootstrapJobArgsLocation The path for the job args file
+   * @param bootstrapAppArgsLocation The path for the app args file
+   * @throws IOException
+   * @return A path to the written driver configuration
+   */
+  String writeDriverConfigurationFile(final String bootstrapJobArgsLocation,
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: Shouldn't this code be in the serializer class?

**Comment by markusweimer** [comment]
I've done a pass.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java (hunk 14 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroMultiRuntimeAppSubmissionParametersSerializer.java]: We write this files on the c# side only. So there is no use for them. I prefer adding them when they will be needed.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java (hunk 16 above)
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapDriverConfigGenerator.java]: No. This is what this class MultiRuntimeYarnBootstrapDriverConfigGenerator does. It uses aother serializers to generate serialize and write the configuration. I don't see a good reason to create a serializer that is just a composition of other serializers.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroYarnJobSubmissionParametersSerializer.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * Reads avro object from file.
+   *
+   * @param file The file to read from
+   * @return Avro object
+   * @throws IOException
+   */
+  static AvroYarnJobSubmissionParameters fromFile(final File file)
+          throws IOException {
```
Comment (shulmanb) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/AvroYarnJobSubmissionParametersSerializer.java]: Done.

**Comment by shulmanb** [comment]
@markusweimer I updated the PR. Please take a look.

File: lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+
+/**
+ * This is a bootstrap launcher for YARN for submission of multiruntime jobs from C#. It allows for Java Driver
+ * configuration generation directly on the Driver without need of Java dependency if REST
+ * submission is used. Note that the name of the class must contain "REEFLauncher" for the time
+ * being in order for the Interop code to discover the class.
+ */
+// TODO[JIRA REEF-1382]: This class does things both for client and driver need to split it
+@Interop(CppFiles = "DriverLauncher.cpp")
+public final class MultiRuntimeYarnBootstrapREEFLauncher {
```
Comment (markusweimer) [inline lang/java/reef-bridge-client/src/main/java/org/apache/reef/bridge/client/MultiRuntimeYarnBootstrapREEFLauncher.java]: Please mark as `@Unstable`, as we plan on doing more work in this area.

**Comment by markusweimer** [comment]
Given that we plan to do more work in the APIs, I changed my code review focus to making sure everything exposed is marked `@Unstable`. I have done a pass to that effect.

**Comment by shulmanb** [comment]
@markusweimer marked as Unstable. LMK is you are ok with merging this and I will do another round of tests with the private c# client I have for this.

**Comment by markusweimer** [comment]
Yes, please test on YARN and comment here how it went.

**Comment by shulmanb** [comment]
@markusweimer @afchung  Ran test on HDI with my test client. Looks good. Please merge.

**Comment by shulmanb** [comment]
@afchung Can you please merge this as well?

**Comment by afchung** [comment]
I'll test and merge.

**Comment by shulmanb** [comment]
@afchung Thanks!
